lib/pure/random.nim

```diff
@@ -516,6 +517,33 @@ proc sample*[T, U](a: openArray[T]; cdf: openArray[U]): T =
     doAssert sample(marbles, cdf) == "blue"
   state.sample(a, cdf)
 
+proc gauss*(r: var Rand; mu = 0.0; sigma = 1.0): float {.since: (1, 3).} =
```

Comment:
```suggestion
proc gauss*(r: var Rand; mu, sigma: float): float {.since: (1, 3).} =
```
Is it better to omit the default values for `mu` and `sigma`?

Reply:
I will let someone else answer that

Reply:
no other defaults would make sense so I'm in favor of keeping `proc gauss*(r: var Rand; mu = 0.0; sigma = 1.0)`

Reply:
The defaults are very nice for an amateur like me who knows what "gauss" means but cannot remember what "mu" and "sigma" mean.

Reply:
> The defaults are very nice for an amateur like me who knows what "gauss" means but cannot remember what "mu" and "sigma" mean.

I don't think we should optimise for this case. Such a person probably shouldn't feel fully confident in their understanding of "mu = 0.0, sigma = 1.0" and would have to read the doc comment, which contains
```
  ## with mean ``mu`` and standard deviation ``sigma``
```
But by that point, they could write it themselves.

Reply:
Can we list the pros and cons? As I see it:

Pros of using default values:
- It's fewer keystrokes for somebody to type `gauss()` than `gauss(0, 1)` or `gauss(0.0, 1.0)`. But I can agree that these default values make more sense than other possibilities. 

Cons:
- It's probably less readable/maintainable  to see `gauss()` than `gauss(0.0, 1.0)`, especially for someone who didn't write the code.
- It allows a user to pass only one parameter and write e.g. `gauss(5.0)`, and the compiler doesn't complain. I argue that this isn't readable - we shouldn't have to remember that the default value is `sigma = 1.0`. And given that it's bad to read, we shouldn't support writing it; a person writing that might be introducing a bug into their program. We can't always protect users from themselves, but we can here.
- I think it's less common/expected. At least Python has [no default values for gauss](https://github.com/python/cpython/blob/a4c09560eaa1e3b9733c7e63d570b6fa7626724c/Lib/random.py#L616).

Can somebody point to a well-used implementation of `gauss` in another language that uses default values for `mu` and `sigma`? Note that I do not assume that something is bad just because it is without precedent. But if default values here are uncommon then it does make it less expected, and therefore harder to read.

Reply:
If there is a probability distribution that justifies having parameter defaults, its a Gaussian -- where the default even has its own name "standard normal distribution"... 

Regarding Python, you rarely use the non-vectorized function you have linked in practice, but rather [`numpy.random.normal`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html?highlight=normal#numpy.random.normal) which does default to standard normal. From my experience that is very convenient to have, and all the examples are perfectly readable in my opinion, just the standard in data science.

Reply:
randn in matlab https://www.mathworks.com/help/matlab/ref/randn.html mu, and sigma is absent
gsl has overloads where sigma = 1 https://www.gnu.org/software/gsl/doc/html/randist.html mu is absent

Reply:
I'm not exactly "against" having the default values - I would also find it convenient to write `gauss()`. I just wanted to ensure that we make the decision consciously.

I think my main objection to default values is what I wrote above: it permits one parameter to be passed. And given that `,` and `.` are adjacent on a lot of keyboards, it's probably guaranteed that people will eventually write e.g.:
- `gauss(100.2)` when they meant `gauss(100,2)`
- `gauss(100.10)` when they meant `gauss(100,10)`

And not be alerted to it by the compiler. Of course, these should really be written with a space after the comma. 

Consider also people who are used to `,` as the decimal separator, or who are implementing something while reading in a language that uses that. Without default values, somebody who writes e.g. `gauss(100,2)` cannot have intended to write `gauss(100.2)` and rely on `sigma = 1`. 

Another possibility for us to consciously reject is for the function to have zero parameters, like in e.g. Go (see [here](https://github.com/golang/go/blob/60f78765022a59725121d3b800268adffe78bde3/src/math/rand/normal.go#L30-L36)). That possibility feels less ergonomic to me.

Reply:
Perhaps my point should be this:

Even if my concerns above are a bit theoretical, is there a reason we can't implement this as something like the below?
- support `gauss()`
- support `gauss(7.0, 2.0)`
- do not support `gauss(7.0)`

or alternatively with `r`:
- support `gauss(r)`
- support `gauss(r, 7.0, 2.0)`
- do not support `gauss(r, 7.0)`

I won't express an opinion here about the design regarding `r`.

Reply:
how about remove the mean parameter? You can only pass one parameter, the std, and your concern is taken care.

Reply:
@b3liever I prefer your original proc signature.

Consider an example: let's say that a user wants to sample from a Gaussian distribution with `mu = 7`, `sigma = 2`. They can write:

```nim
const
  mu = 7
  sigma = 2
```

Now, I think these are our options for what they write next (for simplicity, naming all the procs `gauss` - some are possibly better named `norm`, `randn`, `ugaussian` etc):
1. `let x = gauss(r) * sigma + mu` (which is how e.g. [`normal.go`](https://github.com/golang/go/blob/60f78765022a59725121d3b800268adffe78bde3/src/math/rand/normal.go#L30-L36) does it)
1. `let x = gauss(r, sigma) + mu` (which is how e.g. [`gsl`](https://www.gnu.org/software/gsl/doc/html/randist.html#the-gaussian-distribution) does it. They also provide `ugaussian` which takes only `r` as a parameter, with `sigma = 1`)
1. `let x = gauss(r, mu, sigma)` (which is how e.g. [`numpy.random.normal`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html?highlight=normal#numpy.random.normal) and [`random.py`](https://github.com/python/cpython/blob/a4c09560eaa1e3b9733c7e63d570b6fa7626724c/Lib/random.py#L616)does it)

I think option 3 is the cleanest and the least bug-prone. For example, option 1 allows a user to transpose `mu` and `sigma`, accidentally writing e.g.:
- `let x = gauss(r) * mu + sigma` or
- `let x = mu * gauss(r) + sigma`

My comments above are to say that option 3 is even better if we say:
- both `mu` and `sigma` must be passed
- or neither.

Reply:
> The defaults are very nice for an amateur like me who knows what "gauss" means but cannot remember what "mu" and "sigma" mean.

@Araq Maybe you would prefer the parameters to be named `mean` and `sd`?

Another naming option from above is `loc` and `scale` (see `numpy.random.normal`). But that naming is my least preferred.